language/vm/src/file_format.rs

```diff
@@ -1365,23 +1368,130 @@ impl CompiledScript {
+    dummy_addr: bool,
+    // If an empty signature was added.
+    dummy_sig: bool,
+    // If the <SELF> identifier was added.
+    self_ident: bool,
+    // If a dummy self module handle was added.
+    self_module_handle: bool,
+}
+
 impl CompiledScriptMut {
     /// Converts this instance into `CompiledScript` after verifying it for basic internal
     /// consistency. This includes bounds checks but no others.
+    #[allow(deprecated)]
     pub fn freeze(self) -> Result<CompiledScript, Vec<VMStatus>> {
-        let fake_module = self.into_module();
-        Ok(fake_module.freeze()?.into_script())
+        let (conv_info, fake_module) = self.into_module();
+        Ok(fake_module.freeze()?.into_script(conv_info))
     }
 
     /// Converts a `CompiledScriptMut` to a `CompiledModule` for code that wants a uniform view
     /// of both.
-    pub fn into_module(self) -> CompiledModuleMut {
-        CompiledModuleMut {
+    ///
+    /// This also produces a `ScriptConversionInfo`, which will be required to convert the
+    /// module back into a script.
+    ///
+    /// TODO: rewrite things that depend on this and get this removed.
+    #[deprecated(
+        note = "This function is deprecated and will be removed soon. Please do not introduce new dependencies."
+    )]
+    pub fn into_module(mut self) -> (ScriptConversionInfo, CompiledModuleMut) {
+        // Add the "<SELF>" identifier if it isn't present.
+        //
+        // Note: When adding an element to the table, in theory it is possible for the index
+        // to overflow. This will not be a problem if we get rid of the script/module conversion.
+        let (self_ident, self_ident_idx) = match self
+            .identifiers
+            .iter()
+            .position(|ident| ident.as_ident_str() == self_module_name())
+        {
+            Some(idx) => (false, IdentifierIndex::new(idx as u16)),
+            None => {
+                let idx = IdentifierIndex::new(self.identifiers.len() as u16);
+                self.identifiers
+                    .push(Identifier::new(self_module_name().to_string()).unwrap());
+                (true, idx)
+            }
+        };
+
+        // Use the first module handle as the self module handle. Create one if none exists.
+        //
+        // Note: this correctly handles scripts with either a proper self module handle or no module
+        // handles at all. For scripts whose first module handle is used to refer to external modules,
+        // the verifier may complain about not being able to find implementations for functioins.
+        //
+        // This is because the verifier always considers the first handle self, and it would be very
+        // tricky for us to fix this during conversion. (We don't want to shift the indices.)
+        //
+        // This will no longer be an issue once we remove script/module conversion so this won't be fixed.
+        let (self_module_handle, dummy_addr, self_module_handle_idx) =
```

Comment:
nit `self_module_handle` to `had _self_module_handle` 
nit `dummy_addr` to `needs_dummy_addr`